File: two_sum/two_sum.rb
# def two_sum(nums, target)
# 	num_len = nums.length
# 	i = 0 
#     while i < num_len	
#     	j = i + 1
#         while j < num_len
#             return [i,j] if nums[i] + nums[j] == target
#             j += 1
#         end
#        i += 1
#     end
# end
def two_sum(nums, target)
	num_len = nums.length
	dic = {}
	i = 0 
  while i < num_len	
  	number = nums[i]
  	if number == target / 2 && dic[number]
  		return [dic[number],i]
  	end
  	dic[number] = i
  	maybe = dic[target-number]
  	if maybe && maybe != i
  		return [maybe,i]
  		break
  	end
    i += 1
  end
end
# p two_sum([2, 7, 11, 15],9)
p two_sum([0,4,3,0],0)

File: two_sum/two_sum.java
public class Solution {
    public int[] twoSum(int[] nums, int target) {
    HashMap<Integer, Integer> hmap = new HashMap<Integer, Integer>();
    int num_len = nums.length;
    int i = 0;
    while (i < num_len){
    	int number = nums[i];
    	
    	if (number == target / 2 && hmap.get(number) != null){
    	    return new int[] {hmap.get(number),i};
    	}
    	hmap.put(number,i);
    	if (hmap.get(target-number) != null && hmap.get(target-number) != i){
    	        return new int[] {hmap.get(target-number),i};
    	}
    	i++;
    }
    int[] arr = {0,2}; 
    return arr;
  }
}